subprojects/language-native/src/integTest/groovy/org/gradle/language/AbstractNativeParallelIntegrationTest.groovy

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.language
+
+import org.gradle.api.specs.Spec
+import org.gradle.integtests.fixtures.BuildOperationsFixture
+import org.gradle.internal.execution.ExecuteTaskBuildOperationType
+import org.gradle.internal.operations.trace.BuildOperationRecord
+import org.gradle.nativeplatform.fixtures.AbstractInstalledToolChainIntegrationSpec
+
+abstract class AbstractNativeParallelIntegrationTest extends AbstractInstalledToolChainIntegrationSpec {
+    BuildOperationsFixture buildOperations = new BuildOperationsFixture(executer, temporaryFolder)
+
+    def setup() {
+        executer.withArgument("--max-workers=4")
+    }
+
+    void assertTaskIsParallel(String taskName) {
+        def task = buildOperations.first(ExecuteTaskBuildOperationType, new Spec<BuildOperationRecord>() {
+            @Override
+            boolean isSatisfiedBy(BuildOperationRecord record) {
+                return record.displayName == "Task :${taskName}"
+            }
+        })
+
+        assert task != null
+        assert buildOperations.getOperationsConcurrentWith(ExecuteTaskBuildOperationType, task).size() > 0
+    }
+
+    def withTaskThatRunsParallelWith(String taskName) {
+        buildFile << """
+            import java.util.concurrent.CountDownLatch
+            import org.gradle.internal.work.WorkerLeaseService
+
+            def taskFinished = new CountDownLatch(1)
+            gradle.taskGraph.afterTask { task ->
+                if (task.name == "${taskName}") {
+                    taskFinished.countDown()
+                }
+            }
+            
+            task parallelTask {
+                dependsOn { tasks.${taskName}.taskDependencies }
+                doLast { 
+                    services.get(WorkerLeaseService).withoutProjectLock {
```

Comment:
Shouldn't this be _with_ the project lock, so that we demonstrate that the compile/link task can do its thing while `parallelTask` task is holding the lock? Otherwise we're showing that `parallelTask` can run in parallel with any other task. There's also nothing here that guarantees that `parallelTask` will run at the same time as the compile/link task, for example it might not start until after the task has finished.

What we need is some way to do this:

1. Do not start `parallelTask` until the compile/link task has started.
2. Do not start compiling or linking until `parallelTask` has started (and is holding the lock still).
3. Do not complete `parallelTask` until compiling/linking has completed.

This will show that we did in fact compile/link stuff while another task was running.